File: api/core/menu/product/product.go
package product

import "comies/core/types"

type Product struct {
	ID              types.ID       `json:"id"`
	Code            string         `json:"code"`
	Name            string         `json:"name"`
	Type            types.Type     `json:"type"`
	CostPrice       types.Currency `json:"cost_price"`
	SaleUnit        types.UnitType `json:"sale_unit"`
	MinimumSale     types.Quantity `json:"minimum_sale"`
	MaximumQuantity types.Quantity `json:"maximum_quantity"`
	MinimumQuantity types.Quantity `json:"minimum_quantity"`
	Location        string         `json:"location"`
}

func (p Product) WithID(id types.ID) Product {
	p.ID = id
	return p
}

func (p Product) Validate() (Product, error) {

	if len(p.Code) < 2 || len(p.Code) > 12 {
		return p, ErrInvalidCode
	}

	if len(p.Name) < 2 || len(p.Name) > 60 {
		return p, ErrInvalidName
	}

	if err := ValidateType(p.Type); err != nil {
		return p, err
	}

	if p.CostPrice <= 0 {
		return p, ErrInvalidPrice
	}

	if p.IsOutput() && p.MinimumSale <= types.QuantityMinimum {
		return p, ErrMinimumSaleQuantity
	}

	return p, nil
}

func (p Product) IsComposite() bool {
	return p.Type == InputCompositeType || p.Type == OutputCompositeType
}

func (p Product) IsInput() bool {
	return p.Type == InputCompositeType || p.Type == InputType
}

func (p Product) IsOutput() bool {
	return p.Type == OutputType || p.Type == OutputCompositeType
}

File: ui/src/core/product.ts
export interface Product {
  id?: string;
  code?: string;
  name?: string;
  type?: number;
  cost_price?: number;
  sale_price?: number;
  sale_unit?: string;
  minimum_sale?: number;
  minimum_quantity?: number;
  maximum_quantity?: number;
  location?: string;
}

export namespace Product {
  export const noneType = 0;
  export const outputType = 10;
  export const outputCompositeType = 20;
  export const inputType = 30;
  export const inputCompositeType = 40;

  export const milligram = "mg";
  export const kilogram = "kg";
  export const gram = "g";
  export const unit = "un";

  export const statuses = [
    noneType,
    outputType,
    outputCompositeType,
    inputType,
    inputCompositeType,
  ] as const;

  export const ProductTypeData: { readonly [key: number]: { name: string } } = {
    [noneType]: { name: "Nenhum" },
    [outputType]: { name: "Saída" },
    [outputCompositeType]: { name: "Saída composto" },
    [inputType]: { name: "Entrada" },
    [inputCompositeType]: { name: "Entrada composto" },
  } as const;

  export const ProductUnitData: { readonly [key: string]: { name: string } } = {
    [milligram]: { name: "miligramas" },
    [kilogram]: { name: "quilogramas" },
    [gram]: { name: "grama" },
    [unit]: { name: "unidade" },
  } as const;

  export const isComposite = (t?: number) =>
    t === outputCompositeType || t === inputCompositeType;

  export const isOutput = (t?: number) =>
    t === outputType || t === outputCompositeType;

  export const isInput = (t?: number) => t === inputType || t === inputCompositeType;

  /**
   * clear removes some properties from a product object
   * based on its type
   * @param product The product to be cleaned and analysed
   * @returns The cleaned product object
   */
  export function clear(product: Product) {
    if (isComposite(product.type)) {
      product = {
        ...product,
        location: "",
        minimum_quantity: 0,
        maximum_quantity: 0,
      };
    }

    if (isInput(product.type)) {
      product = { ...product, minimum_sale: 0, sale_price: 0 };
    }

    return product;
  }

  export interface Filter {
    types?: number[]
    code?: string
    name?: string
  }
}
